### PR TITLE
Ignore Spam Txs, Switch RPC if Error, Query RPCs

### PR DESCRIPTION
**Pull request by nick134-bit**
## Description and Motivation

Hey Guys, 
with this PR you can ignore the TX Spam. Just add the Source Addresses to your .env. 
The Bot will add every address that sends a failed transcation and ignore this in the future.
I also added the RPC switcher that catches any timeout errors and prevent crashes.
We also query RPCs from the chainregistry now.

If you have any questions reach out on the White Whale Discord. Channel: Developer/bot-support

## Related Issues

    
#28 #67

---
## Checklist:

- [x] I have read [Migaloo's contribution guidelines](https://github.com/White-Whale-Defi-Platform/migaloo-bots/blob/main/CONTRIBUTING.md).
- [x] My pull request has a sound title and description (not something vague like `Update index.md`)
- [x] All existing and new tests are passing (if any).
- [ ] I updated/added relevant documentation.
- [x] The code is formatted properly using eslint

### CODE REVIEW

File: .env.terra.example [hunk 1]
```diff
@@ -24,7 +24,7 @@ SKIP_BID_RATE="0.1" #e.g. 10% of the profit is used as a bid to win the auction
 BASE_DENOM="uluna"
 GAS_DENOM="uluna"
 CHAIN_PREFIX="terra"
-RPC_URL="" ##change this
```
Comment (kaimen-sano) [inline .env.terra.example:27]: Same here

File: src/core/chainOperator/chainAdapters/cosmjs.ts [hunk 2]
```diff
@@ -155,6 +174,53 @@ class CosmjsAdapter implements ChainOperatorInterface {
 		return mempoolResult.result;
 	}
 
+	/**
+	 * Sets new Clients for Mempoolloop.
+	 *
+	 */
+	public async getNewClients(): Promise<string | void> {
```
Comment (SirTLB) [inline src/core/chainOperator/chainAdapters/cosmjs.ts:181]: So is this function different between cosmjs and injective?

File: src/core/chainOperator/chainoperator.ts [hunk 3]
```diff
@@ -37,21 +37,28 @@ export class ChainOperator {
...
+		//Error handled in getPoolState.
 		return await this.client.queryContractSmart(address, queryMsg);
 	}
 
 	/**
 	 *
 	 */
 	async queryMempool() {
-		return await this.client.queryMempool();
+		try {
```
Comment (SirTLB) [inline src/core/chainOperator/chainoperator.ts:55]: good

File: src/core/chainOperator/chainAdapters/cosmjs.ts [hunk 4]
```diff
@@ -155,6 +174,53 @@ class CosmjsAdapter implements ChainOperatorInterface {
...
+			} else {
+				const errTime = this._timeoutRPCs.get(this._rpcUrls[n]);
+				if (errTime && errTime + TIMEOUTDUR <= currTime) {
+					this._timeoutRPCs.delete(this._rpcUrls[n]);
+					urlString = this._rpcUrls[n];
+				}
+			}
+			n++;
+		}
+		if (!urlString) {
```
Comment (SirTLB) [inline src/core/chainOperator/chainAdapters/cosmjs.ts:202]: shouldnt we add a print in here?

File: src/core/chainOperator/chainAdapters/cosmjs.ts (hunk 2 above)
Comment (SirTLB) [reply src/core/chainOperator/chainAdapters/cosmjs.ts:181]: so we should add this to the chainoperator and only change implementation between the different adapters

File: src/core/types/arbitrageloops/mempoolLoop.ts [hunk 5]
```diff
@@ -92,11 +96,25 @@ export class MempoolLoop {
 				this.totalBytes = +this.mempool.total_bytes;
 			}
 
-			const mempoolTrades: Array<MempoolTrade> = processMempool(this.mempool);
+			const mempooltxs: [Array<MempoolTrade>, Array<{ sender: string; reciever: string }>] = processMempool(
```
Comment (SirTLB) [inline src/core/types/arbitrageloops/mempoolLoop.ts:99]: I still think we need to rework this part a bit

File: src/core/types/base/botConfig.ts [hunk 6]
```diff
@@ -33,6 +35,7 @@ export interface BotConfig {
 	useMempool: boolean;
 	baseDenom: string;
 	gasDenom: string;
+	apiUrl: string | undefined;
```
Comment (SirTLB) [inline src/core/types/base/botConfig.ts:38]: this is too vague of a title

File: src/core/types/base/mempool.ts [hunk 7]
```diff
@@ -60,8 +61,11 @@ export function showTxMemory() {
  *@param mempool The mempool(state) to process.
  *@return An array of swap, send and swap-operation messages that exist in the `mempool`.
  */
-export function processMempool(mempool: Mempool): Array<MempoolTrade> {
-	const mempoolTrades = [];
+export function processMempool(
+	mempool: Mempool,
+	ignoreAddresses: Set<string>,
+): [Array<MempoolTrade>, Array<{ sender: string; reciever: string }>] {
```
Comment (SirTLB) [inline src/core/types/base/mempool.ts:67]: why would we return the sender again in the seond index objects? we have it in MempoolTrade already?

File: .env.juno.example [hunk 8]
```diff
@@ -20,11 +20,15 @@ SKIP_URL= "http://juno-1-api.skip.money"
...
 
+# Addresses to Blacklist. Needed against Spam Txs.
+# For more Info Discord Channel Developers/Bot-Support
+IGNORE_ADDRESSES='[""]'
+
 ##JUNO SETTINGS
 BASE_DENOM="ujuno" # The asset denom to be used as base asset. This should be the denom of a Native Token only.
 GAS_DENOM="ujuno"
 CHAIN_PREFIX="juno"
-RPC_URL="https://juno-rpc.reece.sh" #change this
```
[the remark this answers is not in the document]
Comment (SirTLB) [reply .env.juno.example:27]: can we make it work so it accepts either one rpc url or an array? Im not sure all environments (pm2, npm, docker container or docker compose) support array brackets.

File: src/core/types/base/botConfig.ts [hunk 9]
```diff
@@ -158,3 +177,65 @@ function validateSkipEnvs(envs: NodeJS.ProcessEnv) {
...
+			}
+		}
+	}
+	return out;
+}
+
+/**
+ *
+ */
+async function getRPCfromRegistry(prefix: string, inputurls: any | undefined) {
```
Comment (SirTLB) [inline src/core/types/base/botConfig.ts:216]: no  need to use any here: getRPCfromRegistry(prefix: string, inputurls? : Array<string>){}

File: src/index.ts [hunk 10]
```diff
@@ -41,6 +24,23 @@ startupMessage += "---".repeat(30);
  *
  */
 async function main() {
+	const botConfig = await setBotConfig(process.env);
+
+	let startupMessage = "===".repeat(30);
+	startupMessage += "\n**White Whale Bot**\n";
+	startupMessage += "===".repeat(30);
+	startupMessage += `\nEnvironment Variables:\n
+**RPC ENPDOINT:** \t${botConfig.rpcUrls}
```
Comment (SirTLB) [inline src/index.ts:33]: you need to add logging regarding the setup of the RPCS, how many were entered, how many fetched from chain registry? whats the final list? Same for the ignore addresses. Make sure to log it is used and what the result is.

File: .env.juno.example (hunk 8 above)
[the remark this answers is not in the document]
Comment (kaimen-sano) [reply .env.juno.example:27]: Yep, dotenv does support this syntax.

**Review by SirTLB — APPROVED** [review]

**Pull request merged**